### PR TITLE
Fix error in summary-by-facet for packages

### LINKED ISSUE #1759: Error in `summary-by-facet` Plugin
### Description

Running the following command:
``./scancode -clpieu --json-pp facet_summary.json samples --summary --facet dev=*.java --summary-by-facet``

Generates the following Error
```
ERROR: failed to run post-scan plugin: summary-by-facet:
Traceback (most recent call last):
  File "/home/ayansm/Desktop/GSoD/scancode-toolkit-versions/scancode-toolkit-3.1.1/src/scancode/cli.py", line 1033, in run_codebase_plugins
    plugin.process_codebase(codebase, **kwargs)
  File "/home/ayansm/Desktop/GSoD/scancode-toolkit-versions/scancode-toolkit-3.1.1/src/summarycode/summarizer.py", line 410, in process_codebase
    summarize_codebase_by_facet(codebase, **kwargs)
  File "/home/ayansm/Desktop/GSoD/scancode-toolkit-versions/scancode-toolkit-3.1.1/src/summarycode/summarizer.py", line 452, in summarize_codebase_by_facet
    [(key, sorted_counter(counter)) for key, counter in summary_counters])
  File "/home/ayansm/Desktop/GSoD/scancode-toolkit-versions/scancode-toolkit-3.1.1/src/summarycode/summarizer.py", line 448, in <genexpr>
    for key, values in summarizable_values_by_key.items()
  File "/home/ayansm/Desktop/GSoD/scancode-toolkit-versions/scancode-toolkit-3.1.1/src/summarycode/summarizer.py", line 305, in summarize_values
    return value_summarizers_by_attr[attribute](values)
KeyError: u'packages'

Scanning done.
Some files failed to scan properly:
ERROR: failed to run post-scan plugin: summary-by-facet:
Traceback (most recent call last):
  File "/home/ayansm/Desktop/GSoD/scancode-toolkit-versions/scancode-toolkit-3.1.1/src/scancode/cli.py", line 1033, in run_codebase_plugins
    plugin.process_codebase(codebase, **kwargs)
  File "/home/ayansm/Desktop/GSoD/scancode-toolkit-versions/scancode-toolkit-3.1.1/src/summarycode/summarizer.py", line 410, in process_codebase
    summarize_codebase_by_facet(codebase, **kwargs)
  File "/home/ayansm/Desktop/GSoD/scancode-toolkit-versions/scancode-toolkit-3.1.1/src/summarycode/summarizer.py", line 452, in summarize_codebase_by_facet
    [(key, sorted_counter(counter)) for key, counter in summary_counters])
  File "/home/ayansm/Desktop/GSoD/scancode-toolkit-versions/scancode-toolkit-3.1.1/src/summarycode/summarizer.py", line 448, in <genexpr>
    for key, values in summarizable_values_by_key.items()
  File "/home/ayansm/Desktop/GSoD/scancode-toolkit-versions/scancode-toolkit-3.1.1/src/summarycode/summarizer.py", line 305, in summarize_values
    return value_summarizers_by_attr[attribute](values)
KeyError: u'packages'

```
### System configuration
* What OS are you running on? Ubuntu 16.04 LTS
* What version of scancode-toolkit was used to generate the scan file? v3.1.1
* What installation method was used to install/run scancode? [source download]

### PR DESCRIPTION
**Pull request by inishchith**
Signed-off-by: inishchith <inishchith@gmail.com>

Fixes #1759 

- Add summarize_packages method
- Alter test data to show only package purl

## Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://aboutcode.readthedocs.io/en/latest/scancode-toolkit/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

### CODE REVIEW

**Comment by inishchith** [comment]
@AyanSinhaMahapatra The travis build failure is due to some error in usage of `ref` in Sphinx doc. Can you have a look and let me know how can I fix it [[refer](https://travis-ci.org/nexB/scancode-toolkit/jobs/627621941?utm_medium=notification&utm_source=github_status)]

@pombredanne @JonoYang Can you please have a look and let me know if any changes required.

Thanks!

**Comment by AyanSinhaMahapatra** [comment]
@inishchith It happened because of [this PR](https://github.com/nexB/aboutcode/pull/37) by @majurg. Some links were broken, as intersphinx was not set up for https://scancode-workbench.readthedocs.io. I'll fix this ASAP. Thanks for reporting.

**Comment by inishchith** [comment]
@AyanSinhaMahapatra Thanks for the quick response.

> It happened because of this PR by @majurg. Some links were broken, as intersphinx was not set up for https://scancode-workbench.readthedocs.io.

Oh. I see.

> I'll fix this ASAP. Thanks for reporting.

Sure. You're welcome. 
Great work with the documentation BTW 😅

**Comment by AyanSinhaMahapatra** [comment]
@inishchith Thanks! 

Btw, this issue is being tracked here https://github.com/nexB/scancode-toolkit/issues/1866

**Comment by inishchith** [comment]
@JonoYang @pombredanne Please review this and let me know in case any changes required.

Thanks!

File: src/summarycode/summarizer.py [hunk 1]
```diff
@@ -502,11 +503,29 @@ def package_summarizer(resource, children, keep_details=False):
         logger_debug('package_summarizer: for:', resource,
                      'packages are:', packs)
 
+    package_purls = []
+    for package in packages:        
+        if package.get("purl", False):
+            package_purls.append(package["purl"])
```
Comment (pombredanne) [inline src/summarycode/summarizer.py]: We prefer to use single quotes everywhere unless for docstring... do you mind to update these to use single quote for strings?

File: src/summarycode/summarizer.py [hunk 2]
```diff
@@ -502,11 +503,29 @@ def package_summarizer(resource, children, keep_details=False):
...
+
     # Collect direct children packages summary
     for child in children:
         child_summaries = get_resource_summary(child, key='packages', as_attribute=False) or []
-        packages.extend(child_summaries)
-
+        for child_summary in child_summaries:
+            values = [child_summary['value']] * child_summary['count']
```
Comment (pombredanne) [inline src/summarycode/summarizer.py]: I am a tad puzzled by this line... and I am not sure I get what it does... can you explain a bit? Why create a list with duplicated values like that?

File: src/summarycode/summarizer.py [hunk 3]
```diff
@@ -502,11 +503,29 @@ def package_summarizer(resource, children, keep_details=False):
...
+            package_purls.extend(values)
+ 
     # summarize proper
-    set_resource_summary(resource, key='packages', value=packages, as_attribute=False)
-    return packages
+    packages_counter = summarize_licenses(package_purls)
+    summarized = sorted_counter(packages_counter)
+    set_resource_summary(resource, key="packages", value=summarized, as_attribute=keep_details)
```
Comment (pombredanne) [inline src/summarycode/summarizer.py]: nitpicking again: please use single quotes

File: src/summarycode/summarizer.py [hunk 4]
```diff
@@ -502,11 +503,29 @@ def package_summarizer(resource, children, keep_details=False):
...
+
+    return summarized
+    
+
+def summarize_packages(package_purls):
```
Comment (pombredanne) [inline src/summarycode/summarizer.py]: Please either use `package_urls` or `purls` instead, as it seems clearer to me.

File: src/summarycode/summarizer.py [hunk 5]
```diff
@@ -502,11 +503,29 @@ def package_summarizer(resource, children, keep_details=False):
...
+    """
+    Given a list of package purls, return a mapping of {expression: count
```
Comment (pombredanne) [inline src/summarycode/summarizer.py]: Same comment as above. package urls or purls but package_purls

File: src/summarycode/summarizer.py [hunk 6]
```diff
@@ -502,11 +503,29 @@ def package_summarizer(resource, children, keep_details=False):
...
+    of occurences}
+    """
+    return Counter(package_purls)
```
Comment (pombredanne) [inline src/summarycode/summarizer.py]: This is missing a training line return

File: src/summarycode/summarizer.py (hunk 1 above)
Comment (inishchith) [reply src/summarycode/summarizer.py]: Sure, I'll add it.

File: src/summarycode/summarizer.py (hunk 4 above)
Comment (inishchith) [reply src/summarycode/summarizer.py]: Sorry about this, I'll change these to `package_urls`.

File: src/summarycode/summarizer.py (hunk 6 above)
Comment (inishchith) [reply src/summarycode/summarizer.py]: You mean there should be a blank line before the `return` statement?

Comment (pombredanne) [reply src/summarycode/summarizer.py]: yes, the convention is that all Python sources (and in general most source code text) should end up with a line return. See https://www.reddit.com/r/Python/comments/1zjugg/question_why_does_pep8_recommend_leaving_a_blank/ for some commentary... that a minor nitpicking though! ;)

File: src/summarycode/summarizer.py (hunk 2 above)
Comment (inishchith) [reply src/summarycode/summarizer.py]: @pombredanne Sorry for the delayed response.

I was trying to understand how to collect `package_urls` and count in order to make sure it works as intended. ( ref. @JonoYang 's [comment](https://github.com/nexB/scancode-toolkit/issues/1759#issuecomment-566812536)).

For that reason, I was trying a couple of Ideas :
1. Was this [addition](https://github.com/nexB/scancode-toolkit/pull/1864/files#diff-52211dc043d1ddb236472d7177e24fcaR506)
2. Was to utilize the [existing block](https://github.com/nexB/scancode-toolkit/blob/develop/src/summarycode/summarizer.py#L506) similar to the way`license_summarizer` method does it.

Funny enough, I missed out to remove the other and didn't realize it until you pointed it out.

I'll remove it, Thanks for pointing it out. Please feel free to correct me if I'm wrong.

Thanks!

File: src/summarycode/summarizer.py (hunk 6 above)
Comment (inishchith) [reply src/summarycode/summarizer.py]: Thanks for pointing out. I've made the changes!

**Comment by tdruez** [comment]
> Thank you ++ for this. There are a few comments below for your review.
> @tdruez @JonoYang this would change the way we summarize packages (IMHO for the better and this is alos fixing a bug) ... is that something that would break anything?

I would prefer to postpone any last minute API changes after the upcoming release.

**Comment by inishchith** [comment]
ping! @pombredanne

**Comment by pombredanne** [comment]
@inishchith this is all good and will be merged shortly. I will cut a release before I merge though. Thank you for you patience!

**Comment by inishchith** [comment]
@pombredanne Thanks for the approval ;) 

P.S: I'm looking for more tasks to work on, please do let me know in case you have something specific on top of your mind.

Thanks!

**Comment by pombredanne** [comment]
@inishchith 

> P.S: I'm looking for more tasks to work on, please do let me know in case you have something specific on top of your mind.

At this stage you may want to work on a GSoC proposal if that's part of your plan?

**Comment by inishchith** [comment]
@pombredanne No. I won't be participating in GSoC this year :)

**Comment by pombredanne** [comment]
I am merging this at last.

**Comment by pombredanne** [comment]
This has been merged in #2705 and is in the next release

**Pull request closed without merging**